server/embed/config_test.go

```diff
@@ -743,3 +778,99 @@ func TestUndefinedAutoCompactionModeValidate(t *testing.T) {
+	}{
+		{
+			name: "default",
+			expectedFeatures: map[featuregate.Feature]bool{
+				features.StopGRPCServiceOnDefrag: false,
+				"TestAlpha":                      false,
+				"TestBeta":                       true,
+			},
+		},
+		{
+			name:                                "cannot set experimental flag and feature gate to true at the same time",
+			featureGatesFlag:                    "StopGRPCServiceOnDefrag=true",
+			experimentalStopGRPCServiceOnDefrag: "true",
+			expectErr:                           true,
+		},
+		{
+			name:                                "cannot set experimental flag and feature gate to false at the same time",
+			featureGatesFlag:                    "StopGRPCServiceOnDefrag=false",
+			experimentalStopGRPCServiceOnDefrag: "false",
+			expectErr:                           true,
+		},
+		{
+			name:                                "cannot set experimental flag and feature gate to different values at the same time",
+			featureGatesFlag:                    "StopGRPCServiceOnDefrag=true",
+			experimentalStopGRPCServiceOnDefrag: "false",
+			expectErr:                           true,
+		},
+		{
+			name:                                "can set experimental flag and other feature gates",
+			featureGatesFlag:                    "TestAlpha=true,TestBeta=false",
+			experimentalStopGRPCServiceOnDefrag: "true",
+			expectedFeatures: map[featuregate.Feature]bool{
+				features.StopGRPCServiceOnDefrag: true,
+				"TestAlpha":                      true,
+				"TestBeta":                       false,
+			},
+		},
+		{
+			name:             "can set feature gate",
+			featureGatesFlag: "TestAlpha=true,StopGRPCServiceOnDefrag=true",
+			expectedFeatures: map[featuregate.Feature]bool{
+				features.StopGRPCServiceOnDefrag: true,
+				"TestAlpha":                      true,
+				"TestBeta":                       true,
+			},
+		},
+	}
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			fg := features.NewDefaultServerFeatureGate("test", nil)
+			err := fg.(featuregate.MutableFeatureGate).Add(
+				map[featuregate.Feature]featuregate.FeatureSpec{
+					"TestAlpha": {Default: false, PreRelease: featuregate.Alpha},
+					"TestBeta":  {Default: true, PreRelease: featuregate.Beta},
+				})
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			fg.(featuregate.MutableFeatureGate).Set(tc.featureGatesFlag)
+			getExperimentalFlagVal := func(flagName string) *bool {
+				if flagName != "experimental-stop-grpc-service-on-defrag" || tc.experimentalStopGRPCServiceOnDefrag == "" {
+					return nil
+				}
+				flagVal, parseErr := strconv.ParseBool(tc.experimentalStopGRPCServiceOnDefrag)
+				if parseErr != nil {
+					t.Fatal(parseErr)
+				}
+				return &flagVal
+			}
+			err = SetFeatureGatesFromExperimentalFlags(fg, getExperimentalFlagVal, tc.featureGatesFlag)
+			if tc.expectErr {
+				if err == nil {
+					t.Fatal("expect error")
+				}
+				return
+			}
+			if err != nil {
+				t.Fatal(err)
+			}
```

Comment:
```suggestion			
			if tc.experimentalStopGRPCServiceOnDefrag != "" {
				// overwrite the value for `features.StopGRPCServiceOnDefrag`
				// based on flag `--experimental-stop-grpc-service-on-defrag`.
				getExperimentalFlagVal := func(flagName string) *bool {
					if flagName != "experimental-stop-grpc-service-on-defrag" {
						return nil
					}
					flagVal, parseErr := strconv.ParseBool(tc.experimentalStopGRPCServiceOnDefrag)
					require.NoError(t, parseErr)
					return &flagVal
				}

				err = SetFeatureGatesFromExperimentalFlags(fg, getExperimentalFlagVal, tc.featureGatesFlag)
				if tc.expectErr {
					require.True(t, err != nil, "expect error")
					return
				}
				require.NoError(t, err)
			}			
```

Reply:
> This misses out the test case when the flag is not set.

You can add a separate case for that. Always remember to improve the readability and keep the test case easy to understand

Reply:
personally I don't think putting the function this test is mainly about inside a if block is improving the readability. 
Added some comments, hopefully that helps.

Reply:
1. It isn't about a if block, it's about not coupling a test driving condition `tc.experimentalStopGRPCServiceOnDefrag != ""` with a hard code condition `flagName != "experimental-stop-grpc-service-on-defrag"`. Putting the former in a outer / upper layer gives a clear hint what it's for, so easier to understand.
2. The special missed test case you mentioned above deserves a dedicated test case; do not implicitly test it in this test, otherwise other contributors may update this test in future and miss it again.
3. Use require.NoError to get rid of too many `if err != nil` or `if err == nil`

Reply:
updated code to `require.NoError`